packages/cache/src/service.ts

```diff
@@ -12,19 +12,52 @@ import { validateInputs } from "./utils/validation";
 export class CacheService {
   constructor(private readonly redis: RedisClient) {}
 
+  /**
+   * Wraps Redis operations with connection check and timeout to prevent hanging
+   */
+  private async withTimeout<T>(operation: Promise<T>, timeoutMs = 1000): Promise<T> {
+    return Promise.race([
+      operation,
+      new Promise<T>((_, reject) => {
+        setTimeout(() => {
+          reject(new CacheErrorClass(ErrorCode.RedisOperationError, "Cache operation timeout"));
+        }, timeoutMs);
+      }),
+    ]);
+  }
+
+  /**
+   * Get the underlying Redis client for advanced operations (e.g., Lua scripts)
+   * Use with caution - prefer cache service methods when possible
+   * @returns The Redis client instance or null if not ready
+   */
+  getRedisClient(): RedisClient | null {
+    if (!this.redis.isReady || !this.redis.isOpen) {
+      return null;
+    }
+    return this.redis;
+  }
+
   /**
    * Get a value from cache with automatic JSON deserialization
    * @param key - Cache key to retrieve
    * @returns Result containing parsed value, null if not found, or an error
    */
   async get<T>(key: CacheKey): Promise<Result<T | null, CacheError>> {
+    // Check Redis availability first
+    if (!this.redis.isReady || !this.redis.isOpen) {
```

Comment:
we can extract `!this.redis.isReady || !this.redis.isOpen` into a separate check `isRedisAvailable` (or similar) and then reuse it